In [7]:
import pandas as pd
import numpy as np
from sklearn.model_selection import train_test_split
from sklearn.metrics import accuracy_score, confusion_matrix
from sklearn.linear_model import LinearRegression
from sklearn.metrics import mean_squared_error, r2_score, mean_absolute_error
from sklearn.model_selection import train_test_split
from sklearn.neighbors import KNeighborsRegressor
from sklearn.tree import DecisionTreeRegressor
from sklearn.ensemble import RandomForestRegressor
from sklearn.model_selection import  cross_val_score, KFold


In [8]:
data = pd.read_csv('processed_zomato.csv')

In [3]:
data.head()

Unnamed: 0  type_Café/Bakery  type_Casual Dining  type_Fine Dining  \
0           0               0.0                 0.0               0.0   
1           1               0.0                 0.0               0.0   
2           2               0.0                 0.0               0.0   
3           3               0.0                 0.0               0.0   
4           4               0.0                 1.0               0.0   

   type_Other  type_Quick Bites  city_Ahmedabad  city_Ajmer  city_Alappuzha  \
0         0.0               1.0             0.0         0.0             0.0   
1         0.0               1.0             0.0         0.0             0.0   
2         0.0               1.0             0.0         0.0             0.0   
3         0.0               1.0             0.0         0.0             0.0   
4         0.0               0.0             0.0         0.0             0.0   

   city_Allahabad  ...  delivery  chain_restaurant  indian  asian  western  \
0             0.0  ...       0.0               1.0     1.0    0.0      0.0   
1             0.0  ...       0.0               0.0     1.0    1.0      0.0   
2             0.0  ...       1.0               1.0     0.0    0.0      0.0   
3             0.0  ...       1.0               1.0     1.0    0.0      0.0   
4             0.0  ...       1.0               0.0     1.0    0.0      1.0   

   middle_eastern  fast_food  others  highlights_count  rating  
0             0.0        1.0     1.0          0.333609     4.4  
1             0.0        1.0     0.0          0.300279     4.4  
2             0.0        1.0     0.0          0.500235     4.2  
3             0.0        1.0     1.0          0.422685     4.3  
4             0.0        0.0     0.0          0.765618     4.9  

[5 rows x 112 columns]

In [9]:
data = data.drop('Unnamed: 0', axis=1)

In [22]:
def compare_regression_models(df):
    X = df.drop('rating', axis=1)
    y = df['rating']

    X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, random_state=42)

    
    models = {
        'Decision Tree': DecisionTreeRegressor(random_state=42),
        'Random Forest': RandomForestRegressor(random_state=42),
        'Linear Regression': LinearRegression(),
        'KNN Regressor': KNeighborsRegressor()
   
    kfold = KFold(n_splits=10, shuffle=True, random_state=42)
    results = {}

    for name, model in models.items():
        cv_results = cross_val_score(model, X_train, y_train, cv=kfold, scoring='neg_mean_squared_error')
        results[name] = cv_results

    
    for name, result in results.items():
        print(f'{name}: {result.mean():.4f} ({result.std():.4f})')

    
    best_model = max(results, key=lambda k: np.mean(results[k]))
    models[best_model].fit(X_train, y_train)
    y_pred = models[best_model].predict(X_test)

    mse = mean_squared_error(y_test, y_pred)
    r2 = r2_score(y_test, y_pred)

    print(f'\n{best_model} performance on the test set:')
    print(f'MSE: {mse:.4f}')
    print(f'R^2: {r2:.4f}')

In [27]:
compare_regression_models(data)


Decision Tree: -0.1861 (0.0101)
Random Forest: -0.0963 (0.0069)
Linear Regression: -0.4544 (0.0180)
KNN Regressor: -0.5745 (0.0302)

Random Forest performance on the test set:
MSE: 0.0965
R^2: 0.8995


#### The Random Forest model has the highest mean negative MSE value of -0.0963, which means it has the lowest mean MSE value of 0.0963. As a result, the Random Forest model is selected as the best model.
####  In this case, the R² score of 0.8995 suggests that the Random Forest model explains approximately 89.95% of the variance in the data, which is considered a good performance.